jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressor.scala

```diff
@@ -274,9 +318,18 @@ class XGBoostRegressionModel private[ml] (
             Rabit.init(rabitEnv.asJava)
           }
 
-          val features = batchRow.iterator.map(row => row.getAs[Vector]($(featuresCol)))
-
           import DataUtils._
+          val features = batchRow.iterator.map(row =>
+            if ($(featuresCols).nonEmpty) {
+              // Prefer multiple columns for features
+              val values = row.schema.filter(f => $(featuresCols).contains(f.name))
+                .map(f => row.get(row.schema.fieldIndex(f.name)).toString.toFloat).toArray
+              // All are treated as dense data
+              XGBLabeledPoint(0.0f, null, values)
```

Comment:
IIUC, if users do not want to use vectorAssembler, they treat every row as dense vector?

Reply:
if that's the case, can we filter missing value here to enable the usage of sparse vector?

Reply:
Good findings and good suggestion, However XGBoost already does this to filter missing values from both dense and sparse data, please see [here](https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala#L274). So we don't need to do it again.

Reply:
no, I was saying that

1. because of the usage of vector assembler, xgboost-spark can only accept 0 as the missing value when sparse vector is presented

2. you have only two ways to workaround the limitation in 1 currently, (1) replace all 0s with some impossible values in your dataset; (2) before feed dataframe to xgboost, translate all sparse vector to dense manually.  Both approaches will bring higher memory footprint due to the usage of dense vectors

3. since this PR works on skipping vector assembler, you get some facilities to directly build spark vectors by filtering values (specified as `missing`) other than 0 before feeding them to XGBoost training part

Reply:
Actually this PR has nothing to do the sparse vector i think, but introduces a new training pipeline independently.

1. Current pipeline: **Raw DataFrame => VectorAssembler => XGBoost**.
2. A New pipeline of this PR: **Raw DataFrame => XGBoost**.

I believe it is `VectorAssembler` invovles sparse vector and dense vector. But our "multiple columns for feature" uses the raw feature columns directly, so all of their types are required to be numeric (here is the[ check](https://github.com/dmlc/xgboost/blob/47444e1fb9fa53950e79276726f5d4069e750c9a/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostEstimatorCommon.scala#L39)), not sparse or dense vector. When users choose to use the new pipeline, XGBoost is fitted with the same raw `DataFrame ` as `VectorAssembler `, and build `XGBLabeledPoint`s directly from the raw feature columns and label column. What's mroe, the raw feature values are always "dense". 

As far as I know, there is NO limitation on missing value for dense feature values, then the removing logic [here ](https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala#L274)will works for this new pipeline.

Reply:
I was saying ` Both approaches will bring higher memory footprint due to the usage of dense vectors`

I am repeating again "do NOT force everything to be dense, otherwise, you will have very HIGH memory footprint"

is it still not clear?

Reply:
Thanks but two questions:
1) Is it correct to always remove 0 to save memory for sparse vector? Even this is forced by VectorAssembler.  I mean when 0 is a valid value for feature, removing all the 0s will affect the gradient computation and may build a different model after trainning. ( @trivialfis Please correct me if wrong. ) Since the native will not restore the missing value from the parameter "missing".
2) If it is yes to question1, is it enough to remove 0s to build a sparse vector here? Since missing value will be removed next by [here](https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressor.scala#L290).